internal/core/src/cachinglayer/lrucache/ListNode.cpp

```diff
@@ -0,0 +1,211 @@
+                                  }));
+    };
+    {
+        std::shared_lock<std::shared_mutex> lock(mtx_);
+        if (state_ != State::NOT_LOADED) {
+            return read_op();
+        }
+    }
+    std::unique_lock<std::shared_mutex> lock(mtx_);
+    if (state_ != State::NOT_LOADED) {
+        return read_op();
+    }
+    // need to load.
+    internal::cache_op_result_count_miss(size_.storage_type()).Increment();
+    load_promise_ = std::make_unique<folly::SharedPromise<folly::Unit>>();
+    state_ = State::LOADING;
+    if (dlist_ && !dlist_->reserveMemory(size())) {
+        // if another thread sees LOADING status, the memory reservation has succeeded.
+        state_ = State::ERROR;
+        error_ = folly::make_exception_wrapper<std::runtime_error>(fmt::format(
+            "Failed to load {} due to insufficient resource", key()));
+        load_promise_->setException(error_);
+        load_promise_ = nullptr;
+        return std::make_pair(false, folly::makeSemiFuture<NodePin>(error_));
+    }
+
+    // pin the cell now so that we can avoid taking the lock again in deferValue.
+    auto p = NodePin(this);
+    return std::make_pair(
+        true,
+        load_promise_->getSemiFuture().deferValue(
+            [this, p = std::move(p)](auto&&) mutable { return std::move(p); }));
+}
+
+void
+ListNode::set_error(folly::exception_wrapper error) {
+    std::unique_lock<std::shared_mutex> lock(mtx_);
+    AssertInfo(state_ != State::NOT_LOADED && state_ != State::ERROR,
+               "Programming error: set_error() called on a {} cell",
+               state_to_string(state_));
+    // load failed, release the memory reservation.
+    if (dlist_) {
+        dlist_->releaseMemory(size());
+    }
+    // may be successfully loaded by another thread as a bonus, they will update used memory.
+    if (state_ == State::LOADED) {
+        return;
+    }
+    // else: state_ is LOADING
+    state_ = State::ERROR;
+    load_promise_->setException(error);
+    load_promise_ = nullptr;
+    error_ = std::move(error);
+}
+
+std::string
+ListNode::state_to_string(State state) {
+    switch (state) {
+        case State::NOT_LOADED:
+            return "NOT_LOADED";
+        case State::LOADING:
+            return "LOADING";
+        case State::LOADED:
+            return "LOADED";
+        case State::ERROR:
+            return "ERROR";
+    }
+    throw std::invalid_argument("Invalid state");
+}
+
+void
+ListNode::unpin() {
+    std::unique_lock<std::shared_mutex> lock(mtx_);
+    AssertInfo(
+        state_ == State::LOADED || state_ == State::ERROR,
+        "Programming error: unpin() called on a {} cell, current pin_count {}",
+        state_to_string(state_),
+        pin_count_.load());
+    if (state_ == State::ERROR) {
+        return;
+    }
+    if (pin_count_.fetch_sub(1) == 1) {
+        touch(false);
+    }
+}
+
+void
+ListNode::touch(bool update_used_memory) {
+    auto now = std::chrono::high_resolution_clock::now();
```

Comment:
maintain the access time of list node might be helpful for eviction policy

Reply:
yes we currently maintain the most recent access time. 

we will not update a node's position in DList if it has been updated in the past `dlist_->touch_config().refresh_window` to reduce contention. current default window is 10 seconds. this may cause false eviction, but we should gain more benefits.